### PR TITLE
🎉 New Source: short.io

### PR DESCRIPTION
**Pull request by apostoltego**
## What
New source for [short.io](https://developers.short.io/reference) 

## How
Covering links & raw clicks endpoints

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [ ] Issue acceptance criteria met
- [x ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [x ] Secrets are annotated with `airbyte_secret` in the connector's spec
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [ ] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] `README.md`
    - [ ] `docs/SUMMARY.md` if it's a new connector
    - [ ] Created or updated reference docs in `docs/integrations/<source or destination>/<name>`.
    - [ ] Changelog in the appropriate page in `docs/integrations/...`. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md` contains a reference to the new connector
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by apostoltego** [comment]
@davinchia As discussed on slack here's the PR. 
It should work fine and I managed to get the standard Dockerfile to work on my machine too so it shouldn't be any different. 

Acceptance tests here:

Screenshot 2021-07-29 at 14 08 40

File: airbyte-integrations/connectors/source-shortio/source_shortio/spec.json
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/spec.json:27]: nit: newline

File: airbyte-integrations/connectors/source-shortio/source_shortio/spec.json
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/spec.json:7]: alphabetise

File: airbyte-integrations/connectors/source-shortio/source_shortio/spec.json
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/spec.json:9]: alphabetise

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:266]: nit: newline

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:91]:
```suggestion
        The short.io API can be inconsistent in its inclusion of UTM parameters.
```

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:92]:
```suggestion
        Here, we check if they've been provided and if they haven't, attempt to extract it from the original url.
```

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:60]: are we taking the earliest or latest `createdAt` here?

Comment (davinchia) [reply airbyte-integrations/connectors/source-shortio/source_shortio/source.py:60]: can we rename the variable to make this more readable?

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:68]: this formatting isn't consistent - can you make sure to run `./gradlew format` to make sure the code is formatted?

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:94]:
```suggestion
        utm_response_fields_to_utm_params = {
```

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:105]:
```suggestion
            for resp_field, param in utm_params.items():
```

Comment (davinchia) [reply airbyte-integrations/connectors/source-shortio/source_shortio/source.py:105]: propagate the var name change for readability

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:109]: we already note earlier that we do this because the link doeesn't have all the necessary attributes, so we don't need to repeat it here and can remove this comment

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:112]: can we have a comment explaining what is happening here?

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [reply airbyte-integrations/connectors/source-shortio/source_shortio/source.py:92]: Does this mean there are instances where the UTM params exists in the URL but short.io doesn't return it?

Comment (davinchia) [reply airbyte-integrations/connectors/source-shortio/source_shortio/source.py:92]: btw what does UTM stand for?

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:219]: was this for debugging? 

we should remove this, otherwise it will incur significant performance penalty

File: airbyte-integrations/connectors/source-shortio/source_shortio/schemas/links.json
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/schemas/links.json:151]: nit: newline

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:129]: same comment here on `./gradlew format`

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:263]: instead of passing the entire config here, let's actually have each field as private class vars in each stream, and pass the fields separately. It'll make it easier to tell what field is being used where. Does that make sense?

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:202]: what are 'human' clicks? why does including them prevent unnecessary clicks from coming through?

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/source_shortio/source.py:38]: Can we just use the `TokenAuthenticator` class?

File: airbyte-integrations/connectors/source-shortio/integration_tests/configured_catalog.json
Comment (davinchia) [inline airbyte-integrations/connectors/source-shortio/integration_tests/configured_catalog.json:327]: nit: newline

**Review by davinchia — COMMENTED** [review]
Thanks @apostoltego!

Did an initial pass. I think there are some bits we can clean up to make clear what is happening.

Can I also trouble you to run `./gradlew format` to make sure the code is formatted correctly?

Were you able to run the integration test using your own credentials?

File: airbyte-integrations/connectors/source-shortio/source_shortio/schemas/links.json
Comment (apostoltego) [reply airbyte-integrations/connectors/source-shortio/source_shortio/schemas/links.json:101]: You're right they're not referenced there (same with some other stuff unfortunately the docs aren't complete. Attached is a sample response (from which I've removed the actual response content) as an example of the actual body received. 
Screenshot 2021-08-02 at 13 08 21

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (apostoltego) [reply airbyte-integrations/connectors/source-shortio/source_shortio/source.py:202]: Non human clicks are tagged as bots, crawlers etc from short.io
When using this source we don't want to extract the non-human clicks due to their volume (and them adding no value). 
On our current account 75% of all clicks are bots and these are in the range of millions/month. 

Some resources from short.io:
https://help.short.io/en/articles/4065954-how-short-io-tracks-clicks
https://help.short.io/en/articles/4890644-what-are-the-redirects

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (apostoltego) [reply airbyte-integrations/connectors/source-shortio/source_shortio/source.py:38]: I tried that however the api accepts only the following header
`{
      "Authorization": "KEY"
}`
The `TokenAuthenticator` returns: 
`return {self.auth_header: f"{self.auth_method} {self._token}"}`

The problem is that if you don't pass an `auth_method` or pass `None` this throws an error about leading whitespace. If you pass something like `Bearer XYZ` the API doesn't authenticate. 

If there's an alternative happy to implement

**Comment by apostoltego** [comment]
@davinchia  I've left some comments regarding the issues you've raised. The rest I'll try and resolve asap. 
The connector works without credentials and we're already using our the docker image in our airbyte instance  until this is merged.

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (apostoltego) [reply airbyte-integrations/connectors/source-shortio/source_shortio/source.py:92]: @davinchia correct. Even though their docs reference that they should be present we haven't had them returned consistently. I believe the behaviour depends based on how the link is originally created (whether through their UI or through the API etc. The most common use case for short links is attribution operations (e.g. sharing a link with influencers and including parameters in the original URL in order to trace back purchases and website visits). 

Details on UTM parameters [here](https://en.wikipedia.org/wiki/UTM_parameters)

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [reply airbyte-integrations/connectors/source-shortio/source_shortio/source.py:201]: makes sense! can you add this context to the existing comment?

File: airbyte-integrations/connectors/source-shortio/source_shortio/schemas/links.json
Comment (davinchia) [reply airbyte-integrations/connectors/source-shortio/source_shortio/schemas/links.json:101]: that is wild..

Comment (davinchia) [reply airbyte-integrations/connectors/source-shortio/source_shortio/schemas/links.json:101]: can we create a section in the readme pointing this out? otherwise our python devs are going to be very confused maintaining this :)

File: airbyte-integrations/connectors/source-shortio/source_shortio/source.py
Comment (davinchia) [reply airbyte-integrations/connectors/source-shortio/source_shortio/source.py:37]: Gotcha. What a non-standard way of doing so.

What about something like this:
```
    def get_auth_header(self) -> Mapping[str, Any]:
            return {self.auth_header: f"{self._token}"}
```

**Review by davinchia — APPROVED** [review]
@apostoltego apologies didn't realise the integration tests were already run.

thanks for the quick response. left some replies. I think we are good to merge once all those are addressed.

I've approved for now. Feel free to tag me once everything is addressed and I'll merge it in. Thanks for the contribution!

File: airbyte-integrations/connectors/source-shortio/source_shortio/schemas/links.json
Comment (apostoltego) [reply airbyte-integrations/connectors/source-shortio/source_shortio/schemas/links.json:101]: ofcourse.

**Comment by sherifnada** [comment]
FYI @davinchia sandbox account now available: https://github.com/airbytehq/airbyte/issues/5092#issuecomment-895249898

**Comment by davinchia** [comment]
thanks @sherifnada. been slammed by Cloud - will come back later this week once things are more stable.

**Comment by apostoltego** [comment]
@davinchia  I believe the issues you've raised should be addressed now. Re gradle formatting - to be honest at this point this could be a bit over my head plus I couldn't get everything to work properly on the M1.  I've formatted using black instead. 

Let me know if there's anything else that is required for this to be merged.

**Comment by davinchia** [comment]
Thanks for the effort @apostoltego ! I will follow up with remaining clean up and release tasks

**Pull request merged**
